src/pg_query_summary.c

```diff
@@ -0,0 +1,775 @@
+	ListCell* lc = NULL;
+	foreach(lc, state->cte_names) {
+		char *ctename = lfirst(lc);
+		if (strcmp(ctename, relname) == 0)
+			return true;
+	}
+	return false;
+}
+
+static void handle_from_clause_item(Node* node, ContextType context, WalkState* state) {
+	if (node == NULL)
+		return;
+
+	switch (nodeTag(node)) {
+		case T_JoinExpr: {
+			JoinExpr* je = castNode(JoinExpr, node);
+
+			handle_from_clause_item((Node*) je->larg, context, state);
+			handle_from_clause_item((Node*) je->rarg, context, state);
+
+			handle_from_clause_item(je->quals, context, state);
+
+			break;
+		}
+		case T_RangeVar: {
+			RangeVar* rv = castNode(RangeVar, node);
+			add_table(state, context, castNode(RangeVar, node));
+
+			if (rv->alias) {
+				SummaryAlias* alias = palloc(sizeof(SummaryAlias));
+				alias->key = rv->alias->aliasname;
+				alias->value = range_var_parts_to_name(rv->schemaname, rv->relname);
+				state->aliases = lappend(state->aliases, alias);
+			}
+
+			break;
+		}
+		default:
+			break;
+	}
+}
+
+// Walk the parse tree, storing the results in `summary`.
+//
+// Returns NULL on success.
+PgQueryError* pg_query_summary_walk(Summary* summary, Node* tree) {
+	WalkState state = {0};
+
+	// NOTE regarding cte_names and from_clause_items:
+	// 1. We process cte_names as we iterate through the tree.
+	// 2. We only add items to tables if they are *not* in cte_names.
+	// 3. CTEs can be referenced *after* table names.
+	// 4. Because of this, we have to store from_clause_items and process it after the initial tree walk.
+	//
+	// If we try to do it entirely on the initial pass, we wind up with CTEs being
+	// included in state->tables, and thus in summary->tables.
+
+	pg_query_summary_walk_impl(tree, &state);
+
+	ListCell* lc = NULL;
+	foreach(lc, state.from_clause_items) {
+		NodeWithContext* lc_node_with_context = lfirst(lc);
+
+		Node* lc_node = lc_node_with_context->node;
+
+		handle_from_clause_item(lc_node, lc_node_with_context->context, &state);
+
+		pfree(lc_node_with_context);
+	}
+
+	if (state.tables != NULL)
+		summary->tables = state.tables;
+
+	if (state.aliases != NULL)
+		summary->aliases = state.aliases;
+
+	if (state.cte_names != NULL)
+		summary->cte_names = state.cte_names;
+
+	if (state.functions != NULL)
+		summary->functions = state.functions;
+
+	if (state.filter_columns != NULL)
+		summary->filter_columns = state.filter_columns;
+
+	// No error to return.
+	return NULL;
+}
+
+// Given a walked parse tree and summary, store the truncated version in `summary`.
+//
+// Returns NULL on success.
+static PgQueryError* summary_truncate(Summary* summary, Node* tree)
+{
+	return make_pg_query_error("summary_truncate() is not implemented yet");
+
+	// TODO: implement summary_truncate()
+
+	// No error to return.
+	return NULL;
+}
+
+// Given the summarized query, convert it to protobuf, and store it in `result`.
+static void summary_to_protobuf(PgQuerySummaryParseResult* result, Summary* summary)
```

Comment:
Need to do a quick look over this in terms of memory allocation logic next time.